opensearchtransport/gzip_test.go

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// The OpenSearch Contributors require contributions made to
+// this file be licensed under the Apache-2.0 license or a
+// compatible open source license.
+//
+// Modifications Copyright OpenSearch Contributors. See
+// GitHub history for details.
+
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build !integration
+
+package opensearchtransport
+
+import (
+	"compress/gzip"
+	"io"
+	"strings"
+	"testing"
+)
+
+func TestCompress(t *testing.T) {
+	t.Run("initialize & compress", func(t *testing.T) {
+		gzipCompressor := newGzipCompressor()
+		body := `{"query":{"match_all":{}}}`
+		rc := io.NopCloser(strings.NewReader(body))
+
+		buf, err := gzipCompressor.compress(rc)
+		defer gzipCompressor.collectBuffer(buf)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+
+		// unzip
+		r, _ := gzip.NewReader(buf)
+		s, _ := io.ReadAll(r)
+		if string(s) != body {
+			t.Fatal("expected body to be the same after compressing and decompressing")
+		}
+	})
+
+	t.Run("gzip multiple times", func(t *testing.T) {
+		gzipCompressor := newGzipCompressor()
+		body := `{"query":{"match_all":{}}}`
+		for i := 0; i < 5; i++ {
+			rc := io.NopCloser(strings.NewReader(body))
+
+			buf, err := gzipCompressor.compress(rc)
+			defer gzipCompressor.collectBuffer(buf)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+
+			// unzip
+			r, _ := gzip.NewReader(buf)
+			s, _ := io.ReadAll(r)
+			if string(s) != body {
+				t.Fatal("expected body to be the same after compressing and decompressing")
+			}
+		}
+	})
```

Comment:
If the compressor is a noop and just returns the same string, these tests pass. 

1. Add a test that compressed data is different and smaller than the original data.
2. Add a test that decompressing a random buffer returns the expected errors.
3. Add a test that compresses data, then compresses it again, and decompresses it twice to get to source, ensuring nothing is lost.